[Clang] Overflow Pattern Exclusions

JustinStitt opened the pull request:
## EDIT 08/06

This feature has been renamed to Overflow Pattern Exclusions instead of Overflow Idiom Exclusions.

## Summary
Introduce `-fsanitize-overflow-pattern-exclusion=` which can be used to disable sanitizer instrumentation for common overflow-dependent code patterns.

## Background
For a wide selection of projects, proper overflow sanitization could help catch bugs and solve security vulnerabilities. Unfortunately, in some cases the integer overflow sanitizers are too noisy for their users and are often left disabled. Providing users with a method to disable sanitizer instrumentation of common patterns could mean more projects actually utilize the sanitizers in the first place.

One such project that has opted to not use integer overflow (or truncation) sanitizers is the Linux Kernel. There has been [some discussion](https://lore.kernel.org/all/202404291502.612E0A10@keescook/) recently concerning mitigation strategies for unexpected arithmetic overflow. This discussion is still ongoing and [a succinct article](https://lwn.net/Articles/979747/) accurately sums up the discussion. In summary, many Kernel developers do not want to introduce more arithmetic wrappers when **most developers understand the code patterns as they are**.

Patterns like:
``` c
if (base + offset < base) { ... }
```
or
``` c
while (i--) { ... }
```
or
``` c
#define SOME -1UL
```
… are extremely common in a code base like the Linux Kernel. It is perhaps too much to ask of kernel developers to use arithmetic wrappers in these cases. For example:

``` c
while (wrapping_post_dec(i)) { ... }
```
… which wraps some builtin would not fly. This would incur too many changes to existing code; the code churn would be too much, at least too much to justify turning on overflow sanitizers.

User Cyberax (from [lwn.net](http://lwn.net/)) probably shares a similar question to you:

> I don’t understand why Linux developers won’t just annotate the expected wraparounds. They are used somewhat frequently, but not so much as to affect the readability. Most source code files in the tree won’t have any annotations at all.


but then realizes:

> “I think, the kernel developers just like to be able to read the overflow-dependent code idioms.”

## Prevailing Sentiment

It'd be nice if we could turn on overflow sanitizers for the Linux Kernel (and other projects) without causing too much noise from the compiler. So, we need to disable instrumentation of the noisy code patterns and generally leave the source code intact. This will improve the signal-to-noise ratio and help us find **real** bugs.

## Implementation Details
Currently, this PR tackles three pervasive idioms:

`if (a + b < a)` or some logically-equivalent re-ordering like `if (a > b + a)`

`while (i--)` (for unsigned) a post-decrement always overflows here

`-1UL, -2UL, etc` negation of unsigned constants will always overflow

The patterns that are excluded can be chosen from the following list:
* `add-overflow-test`
* `post-decr-while`
* `negated-unsigned-const`

These can be enabled with a comma-separated list: `-fsanitize-overflow-pattern-exclusion=add-overflow-test,negated-unsigned-const`.

`all` or `none` may also be used to specify that all patterns should be excluded or that none should be.


---

CCs: @efriedma-quic @kees @jyknight @fmayer @vitalybuka

JustinStitt committed:
implement idiom exclusions

Add flag `-fno-sanitize-overflow-idioms` which disables integer
overflow/truncation sanitizer instrumentation for common
overflow-dependent code patterns.

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
add docs

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
format

Signed-off-by: Justin Stitt <justinstitt@google.com>

vitalybuka commented:
I assume you are going to change fronted to avoid compiler abusing these UBs?

vitalybuka commented on clang/docs/UndefinedBehaviorSanitizer.rst:
```diff
@@ -312,6 +312,38 @@ This attribute may not be
+
+.. code-block:: c++
+
+    unsigned long foo = -1UL; // No longer causes a negation overflow warning
```
this are compile time warnings, not runtime warnings?

UBSAN and -fsanitize- is about runtime checks.

JustinStitt replied:
https://godbolt.org/z/r3a31fKvx I am getting runtime errors.

vitalybuka commented:
> I assume you are going to change fronted to avoid compiler abusing these UBs?

Actually stuff like unsigned overflows are not UB.

Maybe instead of  `-fno-sanitize-overflow-idioms` we convert into
`unsigned-integer-overflow`  into ` check group` so user can -fno-sanitize=<special-case-of>-unsigned-integer-overflow

JustinStitt commented:
> Actually stuff like unsigned overflows are not UB.
> 
> Maybe instead of `-fno-sanitize-overflow-idioms` we convert into `unsigned-integer-overflow` into ` check group` so user can -fno-sanitize=-unsigned-integer-overflow

Right, unsigned overflow is well-defined. This isn't about the semantics of "undefined-ness". We want the unsigned sanitizer ON but some of the things it reports are noisy -- that's what this PR is about

JustinStitt commented:
> I assume you are going to change fronted to avoid compiler abusing these UBs?

Well, like you said, none of this is really undefined behavior and isn't getting optimized away. The exception being `if ( a + b < a )` which can definitely have some unexpected results with higher optimization levels. Users can opt into `-fno-strict-overflow` (like the Linux Kernel). I don't think `-fno-sanitize-overflow-idioms` should imply `-fno-strict-overflow`. It's up to the user there, right?

vitalybuka commented:
> Right, unsigned overflow is well-defined. This isn't about the semantics of "undefined-ness". We want the unsigned sanitizer ON but some of the things it reports are noisy -- that's what this PR is about

"Idiom" seems to vague to me. kernel has own opinion on what is idem, other users may have a different.

Also so far we were able to avoid behavior tweaking flags in UBSAN. like `-fno-sanitize-overflow-idioms` (e.g. we have them in Asan). In UBSAN we were able to express as sets of checks. It would be nice to continue these way.

WDYT on splitting existing checks into smaller checks with verbose naming?

kees commented:
`-fsanitize-pattern-exclusion=add-wrap-test,neg-unsigned-const,post-decr-while` ?

JustinStitt commented:
> `-fsanitize-pattern-exclusion=add-wrap-test,neg-unsigned-const,post-decr-while` ?

This looks good to me. I will implement support for comma-separated patterns to be specified.

JustinStitt committed:
add support for comma-separated pattern specification

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented:
@kees @vitalybuka 3ef8d3c implements `-fsanitize=pattern-exclusion=`.

It supports the following values:
`all` - turn all supported pattern exclusions ON
`none` - turn them all off (has precedence over others, I will add a warning soon if users try to use this with "all" or any of the other ones
`add-overflow-test` - excludes the pattern `if (a + b < a)` from sanitizers
`negated-unsigned-const` - excludes the pattern `-5UL` (this needs renaming because it doesn't only work on const)
`post-decr-while` - excludes the pattern `while (i--)` where `i` is unsigned.

I made an effort to rename everything in the source too, not just the user-facing stuff. The tests are updated too.

JustinStitt committed:
remove more instances of 'idiom'

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
default initialize field

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt committed:
update tests to not redefine size_t

Signed-off-by: Justin Stitt <justinstitt@google.com>

bwendling commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -4248,6 +4248,22 @@ bool CompilerInvocation::ParseLangArgs(LangOptions &Opts, ArgList &Args,
+    for (int i = 0, n = A->getNumValues(); i != n; ++i) {
+      StringRef Value = A->getValue(i);
+      if (Value == "none")
+        Opts.OverflowPatternExclusionMask |= LangOptionsBase::None;
```
This seems counter-intuitive. Shouldn't this be an assignment instead?

```c
Opts.OverflowPatternExclusionMask = LangOptionsBase::None;
```

Or is that not the behavior you want?

Also, look up `StringSwitch`, which does the same as these if-then statements, but in the LLVM way. :-)

JustinStitt replied:
resolved with 4b3efbb41ff86eeff15671b1d876e1ef6a58a536

bwendling commented on clang/test/CodeGen/overflow-idiom-exclusion-fp.c:
```diff
@@ -0,0 +1,83 @@
+// Check for potential false positives from patterns that _almost_ match classic overflow-dependent or overflow-prone code patterns
```
In general, I think it's expected the RUN lines will be first and any comments about the test can come afterwards. It allows the comments to be set out more prominently. (Here and the next testcase.)

bwendling commented on clang/docs/UndefinedBehaviorSanitizer.rst:
```diff
@@ -293,6 +293,40 @@ To silence reports from unsigned integer overflow, you can set
+instrumentation for these common patterns one should use
+``-fsanitize-overflow-pattern-exclusion=``.
+
+Currently, this option supports three pervasive overflow-dependent code idioms:
```
It might be useful to add "titles" to each example here. Something like:

```txt
``negated-unsigned-const``

.. code-block:: c++

    /// -fsanitize-overflow-pattern-exclusion=negated-unsigned-const
 ...
```

JustinStitt replied:
thanks. resolved in 4b3efbb41ff86eeff15671b1d876e1ef6a58a536

bwendling commented on clang/docs/UndefinedBehaviorSanitizer.rst:
```diff
@@ -293,6 +293,40 @@ To silence reports from unsigned integer overflow, you can set
+    if (base + offset < base) { /* ... */ } // The pattern of `a + b < a`, and other re-orderings,
+                                            // won't be instrumented (same for signed types)
+
+Negated unsigned constants, post-decrements in a while loop condition and
```
This first line could go to the first paragraph?

JustinStitt replied:
right, that reads better. thanks!

resolved with 4b3efbb41ff86eeff15671b1d876e1ef6a58a536

bwendling commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -555,6 +570,11 @@ class LangOptions : public LangOptionsBase {
   GPUDefaultStreamKind GPUDefaultStream;
 
+  /// Which overflow patterns should be excluded from sanitizer instrumentation
+  int OverflowPatternExclusionMask = 0;
```
Use an `unsigned int` here.

JustinStitt replied:
gotcha, resolved with 4b3efbb41ff86eeff15671b1d876e1ef6a58a536

bwendling commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -2877,6 +2888,17 @@ ScalarExprEmitter::EmitScalarPrePostIncDec(const UnaryOperator *E, LValue LV,
     bool canPerformLossyDemotionCheck = false;
+
+    // Is the pattern "while (i--)" and overflow exclusion?
+    bool disableSanitizer =
```
This is difficult to read. I know it's nice to have it all in one big conditional, but it'd be more readable if split up a bit.

JustinStitt replied:
You're right. I split up the logic into a new method in 4b3efbb41ff86eeff15671b1d876e1ef6a58a536

JustinStitt committed:
separate logic into function and prefer StringSwitch

Addressing Bill's feedback:
* separate while(i--) disableSanitizer logic into a function, rename this
field.

* slightly rewrite some docs by adding titles

* Prefer stringswitch to if-else chain

* move comments within tests to after RUN lines

Signed-off-by: Justin Stitt <justinstitt@google.com>

JustinStitt commented:
@bwendling I tried to address all your feedback points with 4b3efbb41ff86eeff15671b1d876e1ef6a58a536 Let me know how that looks 👍

JustinStitt committed:
remove anonymous namespace

remove the anonymous namespace as a static qualifier is enough for this
method.

Signed-off-by: Justin Stitt <justinstitt@google.com>
---

Thanks Bill.

JustinStitt commented on clang/lib/AST/Expr.cpp:
```diff
@@ -4759,6 +4759,55 @@ ParenListExpr *ParenListExpr::CreateEmpty(const ASTContext &Ctx,
   return new (Mem) ParenListExpr(EmptyShell(), NumExprs);
 }
 
+namespace {
```
resolved by https://github.com/llvm/llvm-project/pull/100272/commits/2e3d4795633bb1a134be10b44e83a025a7b21d8e

thanks

bwendling reviewed (approved):
LGTM

efriedma-quic commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -195,13 +196,23 @@ static bool CanElideOverflowCheck(const ASTContext &Ctx, const BinOpInfo &Op) {
+
+  if (UO && UO->getOpcode() == UO_Minus && UO->isIntegerConstantExpr(Ctx) &&
+      Ctx.getLangOpts().isOverflowPatternExcluded(
+          LangOptions::OverflowPatternExclusionKind::NegUnsignedConst))
```
Check isOverflowPatternExcluded() first, since it's cheaper to check.

JustinStitt replied:
Nice catch, resolved in https://github.com/llvm/llvm-project/pull/100272/commits/cb0fcd1c3b1ded7258bb20cff4d187ca56f17149

efriedma-quic commented on clang/include/clang/AST/Expr.h:
```diff
@@ -3860,6 +3860,7 @@ class CStyleCastExpr final
 class BinaryOperator : public Expr {
   enum { LHS, RHS, END_EXPR };
   Stmt *SubExprs[END_EXPR];
+  bool ExcludedOverflowPattern = false;
```
I'd prefer to avoid increasing the size of BinaryOperator if possible; do we really need to precompute this?  If we do, is there a spare bit in subclass data?

JustinStitt replied:
Ok, I think I found a spare bit in `BinaryOperatorBits` but I am not super familiar with how all these bit classes work.

See: https://github.com/llvm/llvm-project/pull/100272/commits/cb0fcd1c3b1ded7258bb20cff4d187ca56f17149 

Thanks

JustinStitt committed:
move ExcludedOverflowPattern to BinaryOperatorBits

* move ExcludedOverflowPattern to BinaryOperatorBits
* reorder some logic to benefit from short-circuiting

Signed-off-by: Justin Stitt <justinstitt@google.com>

bwendling commented on clang/include/clang/AST/Stmt.h:
```diff
@@ -649,6 +649,8 @@ class alignas(void *) Stmt {
     /// It is 0 otherwise.
     LLVM_PREFERRED_TYPE(bool)
     unsigned HasFPFeatures : 1;
+    LLVM_PREFERRED_TYPE(bool)
```
Nit: Newline and maybe a comment?

JustinStitt replied:
Good idea, see https://github.com/llvm/llvm-project/pull/100272/commits/e8afeed6dfc5d427b83d52d69c8f92b27fd1220e

thanks

JustinStitt committed:
add comment to flag

Signed-off-by: Justin Stitt <justinstitt@google.com>

bwendling commented:
Thanks! @efriedma-quic anymore comments?

efriedma-quic reviewed (commented):
I think serialization is missing for the new bit on BinaryOperator.  I'm not sure why we're storing it in the first place, though; it's queried in exactly one place, so there isn't really any benefit to precomputing it.

JustinStitt commented:
> I think serialization is missing for the new bit on BinaryOperator. 

How do I add this?



> I'm not sure why we're storing it in the first place, though; it's queried in exactly one place, so there isn't really any benefit to precomputing it.

It's queried when we check if we can elide the overflow sanitizer check, at this point we're dealing with the BO responsible for the addition (or subtraction) and not with the BO responsible for comparison. To properly check for the entire pattern I am trying to exclude I am pretty sure I need that comparison BO; so I can either precompute comparison BOs and store a bit on its child BO or I navigate up the parent map from the child during the overflow ellision check.

Which is the better approach?

JustinStitt committed:
add ExcludedOverflowPattern bit to Stmt Serialization

Signed-off-by: Justin Stitt <justinstitt@google.com>
Co-authored-by: Bill Wendling <morbo@google.com>

bwendling commented:
> @efriedma-quic
> 
> > I think serialization is missing for the new bit on BinaryOperator.
> 
> How do I add this?
> 
> > I'm not sure why we're storing it in the first place, though; it's queried in exactly one place, so there isn't really any benefit to precomputing it.
> 
> It's queried when we check if we can elide the overflow sanitizer check, at this point we're dealing with the BO responsible for the addition (or subtraction) and not with the BO responsible for comparison. To properly check for the entire pattern I am trying to exclude I am pretty sure I need that comparison BO; so I can either precompute comparison BOs and store a bit on its child BO or I navigate up the parent map from the child during the overflow ellision check.
> 
> Which is the better approach?

After chatting with Justin, I agree that precomputing this flag is a better option. The flag is set while processing the comparison operator, during which it sets the flag on its children. Trying to go from the bottom-up to see if an addition statement has this flag set is much harder to do and filled with potential pitfalls.

efriedma-quic reviewed (approved):
LGTM

The other thing you could do is modify ScalarExprEmitter::EmitCompare... if it sees an overflow pattern, it passes that down as an argument to the visit of the add.  But I guess given the current structure of the code, that gets very complicated, so just storing the bit is probably easier.

JustinStitt committed:
colocate getter and setter together

previously, the getter was a couple dozen lines about the setter, for no
good reason. Let's put these two methods visually next to one another.

Signed-off-by: Justin Stitt <justinstitt@google.com>

dyung commented:
@bwendling The follow-up commit you made to the tests added here (7275919cd5fc89c42a52168c9f4411b4e5421c95) seems to have caused the test to start failing on at least a few bots.

- https://lab.llvm.org/buildbot/#/builders/154/builds/2752
- https://lab.llvm.org/buildbot/#/builders/133/builds/3174
- https://lab.llvm.org/buildbot/#/builders/56/builds/4889
- https://lab.llvm.org/buildbot/#/builders/190/builds/3745
- https://lab.llvm.org/buildbot/#/builders/144/builds/4656
- https://lab.llvm.org/buildbot/#/builders/46/builds/3267
- https://lab.llvm.org/buildbot/#/builders/60/builds/4994
- https://lab.llvm.org/buildbot/#/builders/73/builds/3786

Can you take a look and revert if you need time to investigate?

JustinStitt commented:
@bwendling we want to use `-triple` not `-target`, can you cherry pick 3ef83ad323dfbe99028c774c3716706671ba7b8b

I also reordered the tests and added labels so the tests are actually checking the right stuff -- debugging was really difficult but will be easier now.

bwendling commented:
Hopefully 6e2d9df02502 fixes the buildbots. I think we should using update_cc_test_checks.py to generate these testcases though.

dyung commented:
> Hopefully [6e2d9df](https://github.com/llvm/llvm-project/commit/6e2d9df02502e16659e4a9397260baf9df224f17) fixes the buildbots. I think we should using update_cc_test_checks.py to generate these testcases though.

The test overflow-idiom-exclusion-fp.c appears to have been fixed, but overflow-idiom-exclusion.c is still failing https://lab.llvm.org/buildbot/#/builders/144/builds/4664.

bwendling commented:
I removed it temporarily.

vitalybuka commented:
We should revert this patch. It breaks mupltiple bots, introducing Msan reports.

Also on a first look I don't like `-fsanitize-pattern-exclusion=` at all, very inconsistent.
I assumed we agreed to splitting sanitizers into smaller sets?
I would like to know why we need to introduce a new approach instead of using existing one.

Sorry, but I am not sure why this didn't show up my https://github.com/llvm/llvm-project/pulls?q=is%3Aopen+is%3Apr+review-requested%3A%40me+sort%3Aupdated-desc I wanted to review this patch.

@MaskRay maybe have opinion here as well?

vitalybuka commented:
Before reland, please include me into review
I'd like to understand why
`-fsanitize-pattern-exclusion=all` is better than something like `-fno-sanitize=overflow-pattern-all`

kees commented:
> Before reland, please include me into review I'd like to understand why `-fsanitize-pattern-exclusion=all` is better than something like `-fno-sanitize=overflow-pattern-all`

The latter doesn't make sense to me. `no-sanitize` takes a list of sanitizers to completely disable. The pattern exclusion option is also listing sanitizers for altering their behavior. I find the option naming in the existing patch intuitive.

JustinStitt commented:
> Sorry, but I am not sure why this didn't show up my https://github.com/llvm/llvm-project/pulls?q=is%3Aopen+is%3Apr+review-requested%3A%40me+sort%3Aupdated-desc I wanted to review this patch.

Did I do something wrong with my PR or fork settings? I am not sure why you weren't notified I left you on the CC and reviewer list. Thanks for your comments -- working on fixing the test cases right now.

vitalybuka commented:
> > Before reland, please include me into review I'd like to understand why `-fsanitize-pattern-exclusion=all` is better than something like `-fno-sanitize=overflow-pattern-all`
> 
> The latter doesn't make sense to me. `no-sanitize` takes a list of sanitizers to completely disable. The pattern exclusion option is also listing sanitizers for altering their behavior. I find the option naming in the existing patch intuitive.

Can we split `-fsanitize=unsigned-integer-overflow` into -fsanitize=unsigned-integer-overflow-patternA,unsigned-integer-overflow-patternB,unsigned-integer-overflow-patternC...' ?

Then it's quite intuitive to disable them with `no-sanitize`.

> > Sorry, but I am not sure why this didn't show up my https://github.com/llvm/llvm-project/pulls?q=is%3Aopen+is%3Apr+review-requested%3A%40me+sort%3Aupdated-desc I wanted to review this patch.
> 
> Did I do something wrong with my PR or fork settings? I am not sure why you weren't notified I left you on the CC and reviewer list. Thanks for your comments -- working on fixing the test cases right now.

I am not sure, maybe you didn't click "re-request review", next to reviewer. I don't really understand GitHub algorithm about hiding from active reviews.

kees commented:
> Can we split `-fsanitize=unsigned-integer-overflow` into `-fsanitize=unsigned-integer-overflow-patternA,unsigned-integer-overflow-patternB,unsigned-integer-overflow-patternC...` ?
> 
> Then it's quite intuitive to disable them with `no-sanitize`.

Yikes, no way. The pattern exclusions apply to all of the overflow-related sanitizers. Each is a small behavioral difference; they're not separate sanitizers nor do they apply to only one sanitizer.

If `-fsanitize-pattern-exclusion` isn't wanted, perhaps `-fsanitize-overflow=pattern-exclude-{none,all,add-overflow-test,post-decr-while,negated-unsigned-const}`? But I still prefer it as-as.

vitalybuka commented:
> > Can we split `-fsanitize=unsigned-integer-overflow` into `-fsanitize=unsigned-integer-overflow-patternA,unsigned-integer-overflow-patternB,unsigned-integer-overflow-patternC...` ?
> > Then it's quite intuitive to disable them with `no-sanitize`.
> 
> Yikes, no way. The pattern exclusions apply to all of the overflow-related sanitizers. Each is a small behavioral difference; they're not separate sanitizers nor do they apply to only one sanitizer.

That's a strong argument, such explanation is good enough to move forward.

> 
> If `-fsanitize-pattern-exclusion` isn't wanted, perhaps `-fsanitize-overflow=pattern-exclude-{none,all,add-overflow-test,post-decr-while,negated-unsigned-const}`? But I still prefer it as-as.

historically we for modifieres we used `-fsanitize-<sanitizer>-<modifier>`, e.g. `-fsanitize-address-outline-instrumentation` or `-fsanitize-address-use-after-return` `-fsanitize-memory-track-origins`

Then `-fsanitize-undefined-exclude-overflow-pattern=none|all|.....`
I slightly prefer `exclude` -> `ignore`

nico commented:
The test added here was tweaked a bit and then deleted in https://github.com/llvm/llvm-project/commit/07a8cbaf8dc16bebf6e875173d20299d9cc47cc5

What gives?

Instead of deleting tests, we should revert the PR that adds them and then reland the PR with working tests.

JustinStitt commented:
@nico 
> The test added here was tweaked a bit and then deleted in [07a8cba](https://github.com/llvm/llvm-project/commit/07a8cbaf8dc16bebf6e875173d20299d9cc47cc5)
> 
> What gives?
> 
> Instead of deleting tests, we should revert the PR that adds them and then reland the PR with working tests.

The tests are back under a new name https://github.com/llvm/llvm-project/pull/104889/files 

And if you're mentioning specifically the tests involving `ZSTD` kernel code, those were removed. Not because they failed or caused build issues but because they are now obsolete and were simply remnants of a time when these pattern exclusions were implemented as an IR optimization pass -- also there was some licensing issues with using un-minimized kernel code in the tests so instead of obfuscating them I just removed them.

tl;dr: there's still tests but I removed one single edge case test that is obsolete.
